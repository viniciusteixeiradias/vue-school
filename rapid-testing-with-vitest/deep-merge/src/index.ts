export const deepMerge = (a, b) => {
  if (Array.isArray(a)) {
    return [...a, ...b]
  }
  
  return Object.assign(a, b)
}