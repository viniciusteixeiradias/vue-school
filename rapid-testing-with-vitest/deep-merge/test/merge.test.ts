import { test, expect } from 'vitest'
import { deepMerge } from '../src'

test('shallow merge', () => {
  const merged = deepMerge(
    {
      name: 'Vinicius'
    },
    {
      github: 'viniciusteixeiradias'
    }
  )
    
  expect(merged).toEqual({
    name: 'Vinicius',
    github: 'viniciusteixeiradias'
  })
})

test('shallow merge with overlaps', () => {
  const merged = deepMerge(
    {
      name: 'Vinicius',
      github: 'unknown'
    },
    {
      github: 'viniciusteixeiradias',
      twitter: 'vinicius'
    }
  )
    
  expect(merged).toEqual({
    name: 'Vinicius',
    github: 'viniciusteixeiradias',
    twitter: 'vinicius'
  })
})

test('shallow merge with arrays', () => {
  const merged = deepMerge(
    ['vue', 'react'],
    ['svelte', 'solid']
  )
    
  expect(merged).toEqual([
    'vue', 'react', 'svelte', 'solid'
  ])
})

test('deep merge with overlaps', () => {
  const merged = deepMerge(
    {
      name: 'Vinicius',
      accounts: {
        github: 'unknown'
      }
    },
    {
      accounts: {
        twitter: 'vinicius'
      }
    }
  )
    
  expect(merged).toEqual({
    name: 'Vinicius',
    accounts: {
      github: 'unknown',
      twitter: 'vinicius'
    }
  })
})

test('deep merge with overlaps snapshot', () => {
  const merged = deepMerge(
    {
      name: 'Vinicius',
      accounts: {
        github: 'unknown'
      },
      languages: ['js']
    },
    {
      accounts: {
        twitter: 'vinicius'
      },
      languages: ['ts', 'vue']
    }
  )
    
  expect(merged).toMatchInlineSnapshot(`
    {
      "accounts": {
        "github": "unknown",
        "twitter": "vinicius",
      },
      "languages": [
        "js",
        "ts",
        "vue",
      ],
      "name": "Vinicius",
    }
  `)
})

test('throws errors on merging two different types', () => {
  expect(() => deepMerge(
    ['foo', 'bar'],
    { foo: 'bar' }
  )).toThrowError('Error: Can not merge two different types')
})