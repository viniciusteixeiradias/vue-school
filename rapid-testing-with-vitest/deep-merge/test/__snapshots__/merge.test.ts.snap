// Vitest Snapshot v1

exports[`deep merge with overlaps snapshot 1`] = `
{
  "accounts": {
    "github": "unknown",
    "twitter": "vinicius",
  },
  "languages": [
    "javascript",
    "typescript",
    "vue",
  ],
  "name": "Vinicius",
}
`;
